File: 백준/Gold/1300. K번째 수/K번째 수.py
N, K = int(input()), int(input())
start, end = 1, K
answer=0

# 10 * 10에서 20보다 작거나 같은 수를 생각해보자.
# 이는 반대로 생각해보면 20을 행으로 나눈 몫이다.
#
# 20//1: 20이지만 열의 개수를 초과할 수 없으니 10개
# 20//2: 10개
# 20//3: 6개


while start <= end:
    mid = (start + end) // 2  # 중간값

    temp = 0
    # 1부터 N행까지 나눠본다.
    for i in range(1, N+1):
        # 20을 행으로 나눈 몫 (최대는 N)
        temp += min(mid//i, N)

    # 이분탐색

    # 갯수가 K보다 크거나 같으면
    if temp >= K:
        answer = mid
        end = mid-1

    # 갯수가 K보다 작으면
    else:
        start = mid+1

print(answer)

File: 백준/Gold/1300. K번째 수/K번째 수.cc
#include <bits/stdc++.h>
using namespace std;

typedef long long ll;

/*
의외로 이분탐색으로 풀 수 있다.

10 * 10 에서 30보다 작거나 같은 수의 개수
: 30을 모든 행으로 나눈 몫의 합

30//1 : 30 => 그러나 행의 개수 10을 넘을 수 없으니 10개
30//2 : 15 => 그러나 행의 개수 10을 넘을 수 없으니 10개
30//3 : 10개
30//4 : 7개
... 30까지 모두 나눈 몫들을 다 더하기

*/

int main(){
    ios_base::sync_with_stdio(false);
    cin.tie(0);
    
    int N, K;
    cin >> N >> K;
    
    int start = 1;
    int end = K;

    int ans;

    while(start<=end){
        int mid = (start+end)/2;

        int rank = 0;
        // 1부터 N행까지 나눈 값 합하기
        for(int row=1; row<=N; row++){
            rank += min(mid/row, N);  // 최대는 행(N)
        }

        // 갯수가 K보다 크거나 같으면 mid 낮추기
        if (rank >= K){
            end = mid-1;
            ans = mid;
        }
        else {
            start = mid+1;
        }
    }
    cout << ans;
    
    return 0;
}